src/class.ts

```diff
@@ -0,0 +1,99 @@
+import {
+    ClassDefinition, PropertyDefinition, ClassDefinitionInput, Expose, ClassPermissions, GroupName,
+} from './types';
+import { createPropertyDefinition } from './property';
+import {
+    EXPOSE_ALL, EXPOSE_EDGE, EXPOSE_NONE, PERMISSIONS,
+} from './constants';
+
+const getRouteName = (name: string, isEdge: boolean) => {
+    if (name.length === 1) {
+        return `/${name.toLowerCase()}`;
+    } if (!isEdge && !name.endsWith('ary') && name.toLowerCase() !== 'evidence') {
+        if (/.*[^aeiou]y$/.exec(name)) {
+            return `/${name.slice(0, name.length - 1)}ies`.toLowerCase();
+        }
+        return `/${name}s`.toLowerCase();
+    }
+    return `/${name.toLowerCase()}`;
+};
+
+const defaultPermissions = (routes: Partial<Expose> = {}): ClassPermissions => {
+    const permissions = {
+        default: PERMISSIONS.NONE,
+        readonly: PERMISSIONS.READ,
+    };
+
+    if (routes.QUERY || routes.GET) {
+        permissions.default |= PERMISSIONS.READ;
+    }
+    if (routes.POST) {
+        permissions.default |= PERMISSIONS.CREATE;
+    }
+    if (routes.PATCH) {
+        permissions.default |= PERMISSIONS.UPDATE;
+    }
+    if (routes.DELETE) {
+        permissions.default |= PERMISSIONS.DELETE;
+    }
+    return permissions;
+};
+
+const createClassDefinition = (opt: ClassDefinitionInput): ClassDefinition => {
+    const { name } = opt;
+    const properties: Record<string, PropertyDefinition> = {};
+
+    for (const propDefn of opt.properties || []) {
+        properties[propDefn.name] = createPropertyDefinition(propDefn);
+    }
+
+    let isEdge = Boolean(opt.isEdge);
+
+    if (opt.targetModel || opt.sourceModel) {
+        isEdge = true;
+    }
+
+    let reverseName;
+
+    if (opt.isEdge) {
+        if (name.endsWith('Of')) {
+            reverseName = `Has${name.slice(0, name.length - 2)}`;
+        } else if (name.endsWith('By')) {
+            reverseName = `${name.slice(0, name.length - 3)}s`;
+        } else if (name === 'Infers') {
+            reverseName = 'InferredBy';
+        } else {
+            reverseName = `${name.slice(0, name.length - 1)}dBy`;
```

Comment:
this line would only be able to handle verbs that end in an e; are there so many verbs it doesn't make sense to just use a dict here?

Reply:
This only needs to reverse the existing edge names which should generally follow the pattern of `<X>By` or `<X>Of` or `<X>s`. These are added by us so we already know the general names. If we add a dict we would probably just be better off hard-coding this in the definition itself. That might be a good idea, and we can definitely do that but I would do it in a separate PR since this is how this code behaved already and this PR is pretty big